### DISCUSSION TITLE
Download opinions for scotus

### DISCUSSION THREAD

**Discussion by raminass** [freelawproject/courtlistener#3019] [General]
what's the best way to get scotus opinions from 1994 until today ? I'm interested in opinions and their plain text. 
Tried the endpoint on the web and filters seems not working.

**Comment by mlissner** [comment]
It's, unfortunately, a bit complex. We used to have simple downloads for this, but generating them became too difficult, and we've since stopped. 

So the way to do this, unfortunately is to either use the bulk data and then filter it down locally (you'll need dockets, clusters, and opinions) or to use the API. 

If you use the API, I think the following query should be helpful:

https://www.courtlistener.com/api/rest/v3/opinions/?cluster__docket__court__id=scotus

That gets you all `Opinion` objects within SCOTUS. 

You said you wanted it during a specific date. The date filed is on the cluster, so:

https://www.courtlistener.com/api/rest/v3/opinions/?cluster__docket__court__id=scotus&cluster__date_filed__gte=1994-01-01

A few things:

1. Note that I'm using double underscores to cross object types. These are like joins in the SQL world.

2. `gte` means Greater Then or Equal to (<kbd>≥</kbd>)

I learned all this by doing http OPTIONS requests on the endpoints:

```
curl -X OPTIONS https://www.courtlistener.com/api/rest/v3/opinions/ | jq
```

For example, that tells us that:

```
    "cluster": {
      "type": "RelatedFilter",
      "lookup_types": "See available filters for 'Opinion Clusters'"
    },
```

That means that you can use `cluster` to get more filters, so I did an OPTIONS request on the cluster, which returned:

```
   "docket": {
      "type": "RelatedFilter",
      "lookup_types": "See available filters for 'Dockets'"
    },
```

That means you can use *dockets* to do more filters. So I did *another* OPTIONS request on the dockets endpoint:

```
curl -X OPTIONS https://www.courtlistener.com/api/rest/v3/dockets/? | jq
```

Which returns (in part):

```
    "court": {
      "type": "RelatedFilter",
      "lookup_types": "See available filters for 'Courts'"
    },
```

Finally, I did an options request on the Court endpoint:

```
curl -X OPTIONS https://www.courtlistener.com/api/rest/v3/courts/? | jq
```

That returned, in part:

```
 "filters": {
    "id": {
      "type": "CharFilter",
      "lookup_types": [
        "exact"
      ]
    },
```

And finally, I put all of that together:

`opinions/?clusters__dockets__courts__id=scotus`

In English:
 - Opinions, that…
 - Have clusters, that…
 - Have dockets, that…
 - Have courts, that…
 - Have an ID of `scotus`.

Whew, that's a tough one! One key to this is that I already *knew* the schema, so I know that we wanted opinions linked to clusters, to dockets, to courts. If I didn't know that, I'd have consulted the schema diagrams in the documentation.

The date is easier, but it's the same thing. The date is on the cluster, so you have to go from opinions, to clusters, to the date.

I hope this helps!
